docs/articles/semantic_chunking.md

```diff
@@ -0,0 +1,537 @@
+
+```python
+from sklearn.metrics import silhouette_score
+def determine_optimal_clusters(embeddings, max_clusters=10):
+    distance_matrix = cosine_distances(embeddings)
+   
+    Z = linkage(distance_matrix, method='average')
+   
+    wcss = []
+    for n_clusters in range(2, max_clusters + 1):
+        cluster_labels = fcluster(Z, t=n_clusters, criterion='maxclust')
+        wcss.append(calculate_wcss(embeddings, cluster_labels))
+   
+    total_variance = np.sum((embeddings - np.mean(embeddings, axis=0))**2)
+    explained_variance = [1 - (w / total_variance) for w in wcss]
+   
+    optimal_clusters = find_elbow_point(range(2, max_clusters + 1), explained_variance)
+   
+    return optimal_clusters
+
+def calculate_wcss(data, labels):
+    n_clusters = len(set(labels))
+    wcss = 0
+    for i in range(n_clusters):
+        cluster_points = data[labels == i+1]
+        cluster_mean = np.mean(cluster_points, axis=0)
+        wcss += np.sum((cluster_points - cluster_mean)**2)
+    return wcss
+
+def find_elbow_point(x, y):
+    diffs = np.diff(y, 2)
+    return x[np.argmax(diffs) + 1]
+
+def chunk_text_with_clusters(text, num_clusters):
+    single_sentences_list = _split_sentences(text)
+    combined_sentences = _combine_sentences(single_sentences_list)
+    embeddings = get_embeddings(combined_sentences)
+
+    distance_matrix = cosine_distances(embeddings)
+    Z = linkage(distance_matrix, method='average')
+    cluster_labels = fcluster(Z, t=num_clusters, criterion='maxclust')
+
+    chunks = []
+    current_chunk = []
+    current_label = cluster_labels[0]
+
+    for i, sentence in enumerate(single_sentences_list):
+        if cluster_labels[i] == current_label:
+            current_chunk.append(sentence)
+        else:
+            chunks.append(' '.join(current_chunk))
+            current_chunk = [sentence]
+            current_label = cluster_labels[i]
+
+    if current_chunk:
+        chunks.append(' '.join(current_chunk))
+
+    return chunks
+
+
+def chunk_text(text, max_clusters=10):
+    single_sentences_list = _split_sentences(text)
+    combined_sentences = _combine_sentences(single_sentences_list)
+    embeddings = get_embeddings(combined_sentences)
+   
+    optimal_clusters = determine_optimal_clusters(embeddings, max_clusters)
+   
+    return chunk_text_with_clusters(text, num_clusters=optimal_clusters)
+```
+
+You can change the value of max_clusters depending on the length of the text. Let us plot the final results. The elbow point is where the rate of decrease sharply shifts.
+
+![Optimal clusters calculation](../assets/use_cases/semantic_chunking/graph3.png)
+
+Silhouette score considers both how similar an object is to its cluster and how different it is from other clusters. A high silhouette score indicates that the clustering is appropriate, while a low score might suggest that the number of clusters needs to be reconsidered.
+

```

Comment:
@Ashish-Abraham
 5 clusters is an optimal cutoff point then, and should give us the most internally similar and externally distinct clusters? 
 Or is it 4 clusters?

Reply:
5 is the optimal cutoff calculated by the algorithm. 4 was a value I randomly tried for illustration.